### PR TITLE
methods for growing sparse matrix in raws and elements separately

### PR DESCRIPTION
**Pull request by SanSanch5**
Signed-off-by: Aleksandr Pakhomov <a.a.pakhomoff@gmail.com>

### CODE REVIEW

File: NeoML/src/TraditionalML/SparseFloatMatrix.cpp [hunk 1]
```diff
@@ -108,6 +108,40 @@ CSparseFloatMatrix& CSparseFloatMatrix::operator = ( const CSparseFloatMatrix& v
...
+{
+	assert( newRowsBufferSize > 0 );
+	if( newRowsBufferSize > body->RowsBufferSize ) {
+		int newBufferSize = max( body->RowsBufferSize * 3 / 2, newRowsBufferSize );
+		CPtr<CSparseFloatMatrixBody> newBodyPart = FINE_DEBUG_NEW CSparseFloatMatrixBody( 0, 0, 0, newBufferSize, 0 );
+
+		::memcpy( newBodyPart->Desc.PointerB, body->Desc.PointerB, body->Desc.Height * sizeof( int ) );
+		::memcpy( newBodyPart->Desc.PointerE, body->Desc.PointerE, body->Desc.Height * sizeof( float ) );
+
+		CSparseFloatMatrixBody* modifiableBody = body.CopyOnWrite();
```
Comment (SAngeliuk) [inline NeoML/src/TraditionalML/SparseFloatMatrix.cpp]: We already have a body (newBodyPart). Why do you get additional?

Comment (SanSanch5) [reply NeoML/src/TraditionalML/SparseFloatMatrix.cpp]: It's not additional, it's current modifiable version. I'm using newBodyPart as a wrapper only to allocate new particular buffers and then deallocate the current old ones (that's why I swap them)

File: NeoML/src/TraditionalML/SparseFloatMatrix.cpp [hunk 2]
```diff
@@ -27,8 +27,8 @@ inline CSparseFloatMatrix::CSparseFloatMatrixBody* CSparseFloatMatrix::CSparseFl
 	CSparseFloatMatrixBody* body = FINE_DEBUG_NEW CSparseFloatMatrixBody( Desc.Height, Desc.Width, ElementCount, RowsBufferSize, ElementsBufferSize );
 	::memcpy( body->Desc.Columns, Desc.Columns, ElementsBufferSize * sizeof( int ) );
 	::memcpy( body->Desc.Values, Desc.Values, ElementsBufferSize * sizeof( float ) );
-	::memcpy( body->Desc.PointerB, Desc.PointerB, Desc.Height * sizeof( int ) );
-	::memcpy( body->Desc.PointerE, Desc.PointerE, Desc.Height * sizeof( float ) );
+	::memcpy( body->Desc.PointerB, Desc.PointerB, RowsBufferSize * sizeof( int ) );
```
Comment (SAngeliuk) [inline NeoML/src/TraditionalML/SparseFloatMatrix.cpp]: RowsBufferSize -> Desc.Height ?!

File: NeoML/src/TraditionalML/SparseFloatMatrix.cpp [hunk 3]
```diff
@@ -57,27 +63,21 @@ CSparseFloatMatrix::CSparseFloatMatrixBody::CSparseFloatMatrixBody( const CSpars
...
-	if( ElementsBufferSize > 0 ) {
-		CurrentMemoryManager::Free( Desc.Columns );
-		CurrentMemoryManager::Free( Desc.Values );
-	}
+	ColumnsBuf.SetSize( ElementsBufferSize );
+	ValuesBuf.SetSize( ElementsBufferSize );
+	BeginPointersBuf.SetSize( RowsBufferSize );
+	EndPointersBuf.SetSize( RowsBufferSize );
+
+	::memcpy( ColumnsBuf.GetPtr(), desc.Columns, ElementsBufferSize * sizeof( int ) );
```
Comment (SAngeliuk) [inline NeoML/src/TraditionalML/SparseFloatMatrix.cpp:72]: ElementCount?

**Review by SAngeliuk — APPROVED** [review]

**Pull request merged**